lib/swig/SharedLibraryModel.h

```diff
@@ -0,0 +1,322 @@
+// SharedLibraryModel
+//----------------------------------------------------------------------------
+template<typename scalar = float>
+class SharedLibraryModel
+{
+public:
+    //----------------------------------------------------------------------------
+    // Typedefines
+    //----------------------------------------------------------------------------
+    typedef void (*VoidFunction)(void);
+
+    typedef std::array< VoidFunction, 5 > VoidIOFuncs;
+    typedef std::tuple< VoidIOFuncs, VoidIOFuncs > PopulationIO;
+    typedef std::unordered_map< std::string, PopulationIO > PopIOMap;
+
+
+    SharedLibraryModel() : m_Library(nullptr), m_AllocateMem(nullptr),
+        m_Initialize(nullptr), m_InitializeModel(nullptr),
+        m_StepTimeGPU(nullptr), m_StepTimeCPU(nullptr)
+    {
+    }
+
+    SharedLibraryModel(const std::string &pathToModel, const std::string &modelName)
+    {
+        if(!open(pathToModel, modelName)) {
+            throw std::runtime_error("Unable to open library");
+        }
+    }
+
+    ~SharedLibraryModel()
+    {
+        // Close model library if loaded successfully
+        if(m_Library)
+        {
+            exitGeNN();
+#ifdef _WIN32
+            FreeLibrary(m_Library);
+#else
+            dlclose(m_Library);
+#endif
+        }
+    }
+
+    //----------------------------------------------------------------------------
+    // Public API
+    //----------------------------------------------------------------------------
+    bool open(const std::string &pathToModel, const std::string &modelName)
+    {
+#ifdef _WIN32
+        const std::string libraryName = pathToModel + modelName + "_CODE\\runner.dll";
+        m_Library = LoadLibrary(libraryName.c_str());
+#else
+        const std::string libraryName = pathToModel + modelName + "_CODE/librunner.so";
+        m_Library = dlopen(libraryName.c_str(), RTLD_NOW);
+#endif
+
+        // If it fails throw
+        if(m_Library != nullptr) {
+            m_AllocateMem = (VoidFunction)getSymbol("allocateMem");
+            m_Initialize = (VoidFunction)getSymbol("initialize");
+            m_InitializeModel = (VoidFunction)getSymbol("init" + modelName);
+
+            m_StepTimeCPU = (VoidFunction)getSymbol("stepTimeCPU", true);
+            m_StepTimeGPU = (VoidFunction)getSymbol("stepTimeGPU", true);
+
+            m_ExitGeNN = (VoidFunction)getSymbol("exitGeNN");
+
+            return true;
+        }
+        else {
+#ifdef _WIN32
+            std::cerr << "Unable to load library - error:" << std::to_string(GetLastError()) << std::endl;;
+#else
+            std::cerr << "Unable to load library - error:" << dlerror() << std::endl;
+#endif
+            return false;
+        }
+
+    }
+
+    // Retrive symbols to pull/push from/to the device from shared model
+    // and store them in a map for fast lookup
+    void initIO( const std::string &popName,
+                 const std::array<bool, 5> &availableDTypes )
```

Comment:
It is super not-important but you could potentially use a ``std::bitset<5>``  here

Reply:
:+1: